### PR TITLE
[Edit] Added Save Dropdown Menu

### PR DESCRIPTION
**Pull request by BogdanAlexandru**
Adds a dropdown menu showing the _Save and Continue Editing_ and _Save and Done Editing_ options at edit-time; it solves #588 

See it in action: http://i.imgur.com/AFw8zOi.gifv

## Author Checklist

1. Changes address original issue? Y - But this still requires an icon for the "Save and Continue Editing" option; right now it uses the regular save icon.
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Changes have been smoke-tested? Y

### CODE REVIEW

**Comment by BogdanAlexandru** [comment]
If all goes smoothly the UI changes will arrive as a separate PR this weekend.

**Comment by VWoeltjen** [comment]
Changes look good to me. @akhenry - would you mind taking a look as well? Given that this is a slightly larger changeset it would be useful to have a second set of eyes.

@BogdanAlexandru, do you plan to make the corresponding UI changes on this PR? This is a sensible refactor toward #588 on its own, so I think it would make sense to merge as an incremental change, but want to make sure we're on the same page first.

**Comment by BogdanAlexandru** [comment]
@VWoeltjen I'd rather get this in on its own. My plan right now is to get the UI related stuff in a PR this weekend; I have a feeling that one will lead to a few changes down the review line so keeping it separate is probably the best approach: it won't make this one's size grow too big & be hard to review.

**Comment by akhenry** [comment]
@VWoeltjen Change set looks good to me.

@charlesh88 As well as "Save" there is "Save As". Save As currently exits the editor. Did you envisage there being an option to continue editing after save as? If so, would this be an option from the editor, or from the save dialog that pops up?

**Comment by BogdanAlexandru** [comment]
On second thought, since these changes have already been reviewed I think going with separate PRs would cause more trouble than it's worth. I'll just push the UI changes on this branch too.

**Comment by akhenry** [comment]
@BogdanAlexandru Ooh, good catch. That didn't occur to me in the review.

I have created a separate issue for dealing with save options for Save As - #1181. For now, so long as new bugs are not introduced to Save As, it's fine that it retains its previous behavior.

**Comment by BogdanAlexandru** [comment]
It's my omission, I was so focused on Save I didn't think about checking the surroundings at all. I'll get this through some serious testing right now (which I should've done in the first place).

This code change was completely immune to unit tests; the real objects' interface changed, but the mocks remained like before, so tests for users of the changed objects kept passing unless I was aware and changed them beforehand. 

How would you feel about adding some integration tests for the critical bits, like the Save As action for example? Something simple resembling a smoke test, that essentially uses mocks only for the unimportant parts and resembles:

```
// Sanity check: if these are not met, it becomes impossible to edit anything
// unless we restart the session.
saveAsAction.perform()
then:
assert(objectUnderEditNotBeingEditedAnymore)
assert(transactionServiceHasNoActiveTransactionAnymore)
```

Since we're using dependency injection everywhere via require.js it should be pretty easy to integrate only what we need and mock what we don't.

**Comment by BogdanAlexandru** [comment]
Bugs seem squashed, all users of EditorCapability should be updated now.

**Comment by BogdanAlexandru** [comment]
Whoops, I wanted to cancel writing the comment and instinctively went for the button next to Comment but that closed the PR. Reopened, sorry for the noise.

Bugs should be squashed, all users of EditorCapability should be updated now. I'll go through these last few changes again tomorrow to make sure I didn't miss any details before moving on to either integration tests if you are okay with them or pushing the UI changes.

**Comment by akhenry** [comment]
@BogdanAlexandru 
Re: integration tests, great question. Our unit tests failed to catch a problem here because they operate in isolation, and an upstream implementation detail affected downstream behavior that we weren't testing for. It's useful food for thought for the team when considering our testing strategy. We accept that unit tests aren't going to catch everything though, which is why we have manual testing as well at the end of each sprint. In this case, the aberrant behavior would have quickly become obvious in our end-of-sprint testing.

TLDR; we don't expect code that is merged into master to be completely bug free.

In the interests of getting this merged, let's not break strategy here. I think it's sufficient to just add a test to `SaveAsActionSpec` that verifies that `EditorCapability.finish()` is called. The `EditorCapabilitySpec` [already verifies]((https://github.com/nasa/openmct/pull/1170/files#diff-d6976288f78af59ec7db05a5b018b7eeL148) ) that `EditorCapability.finish()` cancels the transaction.

#1183 is fixed now, so if you rebase over, or merge from master it will fix the build error.

**Comment by BogdanAlexandru** [comment]
@akhenry 
Ah, alright. I was unaware that's the methodology here. Got it.

Here's the new save menu in action (slow-ish gif): http://i.imgur.com/AFw8zOi.gifv
All that's missing is an icon for the continue option. Do we add it as part of this PR or later?

I'll push my current changes for review in a bit.

**Comment by BogdanAlexandru** [comment]
Pushed the latest commits. The __EditActionControllerSpec__ has been heavily reworked; I initially tried to use jasmine only for mocking save actions but since I needed to pinpoint specific save actions for menu converter correctness checks, it proved to result in a messy mess. So I wrote a bunch of non-jasmine mocked actions and prefixed them with Fake to make it easier to tell they're not jasmine-driven.

_"uses a click handler to perform the clicked action"_ doesn't really need to use the Fake save actions, but it uses them for consistency's sake.

**Comment by BogdanAlexandru** [comment]
Renaming SaveAs' key from "save" to "save-as" in the dropdown menu commit made CreateAction not work anymore for things like Web Page. I actually renamed it for clarity, if it was intended to have the same key as the  regular "Save" action, let me know.

**Comment by BogdanAlexandru** [comment]
> I initially tried to use jasmine only for mocking save actions but since I needed to pinpoint specific save actions for menu converter correctness checks, it proved to result in a messy mess.

Disregard that, I must've not thought about it enough. Went with spy objects in the latest commit and removed the cruft of the extra fakes. Looking back I've no idea why I didn't use this in the first place.

File: platform/commonUI/edit/test/actions/SaveAsActionSpec.js [hunk 1]
```diff
@@ -156,6 +156,28 @@ define(
                 expect(SaveAsAction.appliesTo(actionContext)).toBe(false);
             });
 
+            it("uses the editor capability to save the object", function () {
+                mockEditorCapability.save.andReturn(new Promise(function () {}));
```
Comment (akhenry) [inline platform/commonUI/edit/test/actions/SaveAsActionSpec.js]: We tend to use [mock promises](https://github.com/BogdanAlexandru/openmct/blob/2a2e9ef99df364a9b9c2780178caff40c174833f/platform/commonUI/edit/test/actions/SaveAsActionSpec.js#L43) in situations like this. They are promises that are resolved immediately upon a call to `.then()`, which allows asynchronous code to be tested synchronously. I'd rather not add calls to `runs` and `waitsFor` as these have been deprecated in Jasmine 2.

File: platform/commonUI/edit/bundle.js [hunk 2]
```diff
@@ -203,20 +205,30 @@ define([
                     ]
                 },
                 {
+                    "key": "save-and-stop-editing",
+                    "category": "save",
+                    "implementation": SaveAndStopEditingAction,
+                    "name": "Save and Done Editing",
```
Comment (charlesh88) [inline platform/commonUI/edit/bundle.js]: Save and Finish Editing sounds like you'll go back and "finish" editing. Save and Done is awkward, admittedly, but I think is clearer and still brief.

Comment (BogdanAlexandru) [inline platform/commonUI/edit/bundle.js]: I think Finish sounds better than Done in this context, yup.

File: platform/commonUI/edit/res/templates/edit-action-buttons.html [hunk 3]
```diff
@@ -20,11 +20,30 @@
...
+                     structure="{
+                        options: saveActionsAsMenuOptions,
+                        click: saveActionMenuClickHandler,
+                        cssclass: 'btn-bar right icon-save no-label major'
+                     }">
+        </mct-control>
+    </span>
+
+    <span ng-repeat="currentAction in otherEditActions">
         <a class='s-button {{currentAction.getMetadata().cssclass}}'
```
Comment (BogdanAlexandru) [inline platform/commonUI/edit/res/templates/edit-action-buttons.html]: I could try to get these to be mct-controls too, it shouldn't cause trouble and that way we'd have the whole file be consistent.

File: platform/commonUI/edit/bundle.js (hunk 2 above)
Comment (akhenry) [inline platform/commonUI/edit/bundle.js]: @BogdanAlexandru Ignore what I said here. Save and Done is fine. 😃

File: platform/commonUI/edit/src/capabilities/EditorCapability.js [hunk 4]
```diff
@@ -81,29 +81,28 @@ define(
...
+         * Finish the current editing session. This will discard any pending
          * persist operations
          * @returns {*}
          */
-        EditorCapability.prototype.cancel = function () {
+        EditorCapability.prototype.finish = function () {
             var domainObject = this.domainObject;
             return this.transactionService.cancel().then(function () {
                 domainObject.getCapability("status").set("editing", false);
-                return domainObject;
```
Comment (BogdanAlexandru) [inline platform/commonUI/edit/src/capabilities/EditorCapability.js]: I've got to look back to be sure but I think it stopped being useful once Save was changed.

File: platform/commonUI/edit/res/templates/edit-action-buttons.html (hunk 3 above)
Comment (BogdanAlexandru) [inline platform/commonUI/edit/res/templates/edit-action-buttons.html]: `CancelAction`'s CSS includes `no-label` but I have to pass `metadata.name` as text for each action, and `mct-button` thinks the control needs a label, so it adds `labeled` to its CSS. A label doesn't show but the cancel button appears slightly wider than normal. It cannot be unseen.

I can solve this by making CancelAction's name `undefined`. Since the action appears to only be relevant in the edit bar, I don't think this will result in any control missing a label saying "Cancel". It'd raise the eyebrows of anyone reading the code though.

I'm leaning towards leaving the `otherEditActions` as they were, especially since it gets rid of the nasty `cssclass: ($index === 0 && saveActions.length === 0 ? 'major ' : '') + currentAction.getMetadata().cssclass`. Which makes the first `otherEditActions` button major.

File: platform/commonUI/edit/bundle.js (hunk 2 above)
Comment (charlesh88) [inline platform/commonUI/edit/bundle.js]: Actually, Pete has convinced me that Save and Finish is better than Save and Done. And, we can always tweak this later if needed. :)

File: platform/commonUI/edit/test/actions/SaveAsActionSpec.js (hunk 1 above)
Comment (BogdanAlexandru) [inline platform/commonUI/edit/test/actions/SaveAsActionSpec.js]: I think every relevant dependency already has its promises mocked; the reason `waitsFor` was needed is the same as this:

https://github.com/nasa/openmct/commit/86fcf190660bfd0bbd2d965b031c05c09bfe43e6#diff-696b8ae07d16ac15f27ed229404497ea

Slightly unrelated but dropping it here since it's a small change that I could do as part of this PR: `SaveAsAction` gets a `CreationService` as dependency but it doesn't seem to do anything with it (not even pass it to something else). It's been like this since its inception https://github.com/nasa/openmct/commit/26e368f52dc071e17d7ed8dba79834bc53d8f6bd#diff-55181241cfa7aebea9080d84a4490c0b

Can the dependency be removed, or am I missing something?

File: platform/commonUI/edit/bundle.js (hunk 2 above)
Comment (akhenry) [inline platform/commonUI/edit/bundle.js]: lol. 

@BogdanAlexandru The consensus is "Save and Finish" 😃 Sorry for being so collectively indecisive.

File: platform/commonUI/edit/res/templates/edit-action-buttons.html (hunk 3 above)
Comment (akhenry) [inline platform/commonUI/edit/res/templates/edit-action-buttons.html]: Yeah, just leave them as they are, we'll get them on a future pass. Thanks anyway.

Comment (akhenry) [inline platform/commonUI/edit/res/templates/edit-action-buttons.html]: Created a [new issue](https://github.com/nasa/openmct/issues/1198) to fix this later.

File: platform/commonUI/edit/src/capabilities/EditorCapability.js (hunk 4 above)
Comment (akhenry) [inline platform/commonUI/edit/src/capabilities/EditorCapability.js]: If there's not a strong reason to remove it I'd be inclined to leave it as it was, just to mitigate risk a bit.

Comment (BogdanAlexandru) [inline platform/commonUI/edit/src/capabilities/EditorCapability.js]: Gotcha, the change has been reverted. I removed it initially because it didn't appear to be used anymore at least after a few searches.

File: platform/commonUI/edit/test/actions/SaveAsActionSpec.js (hunk 1 above)
Comment (BogdanAlexandru) [inline platform/commonUI/edit/test/actions/SaveAsActionSpec.js]: Also I noticed that we're using Jasmine 2.3; should I update SaveAsSpec to use the new async testing functionality? http://jasmine.github.io/2.0/introduction.html#section-Asynchronous_Support

Comment (akhenry) [inline platform/commonUI/edit/test/actions/SaveAsActionSpec.js]: @BogdanAlexandru Let me get back to you re: `waitsFor`. re: CreationService, go ahead and remove it. It's a hangover from refactoring of the `SaveAction` and is obsolete.

Comment (akhenry) [inline platform/commonUI/edit/test/actions/SaveAsActionSpec.js]: re: Jasmine 2.3, that dependency is deceptive. The tests are run by karma, and karma-jasmine uses its own version of Jasmine, which is 1.3.1 by the looks of it.

Ok, for now, let's stick with `waitsFor`. The [migration path](http://jasmine.github.io/2.0/upgrading.html#section-Asynchronous_Specs) for these async constructs is well defined, and straightforward.

Comment (BogdanAlexandru) [inline platform/commonUI/edit/test/actions/SaveAsActionSpec.js]: Ah alright, that definitely confused me then. Pushed the latest commit w/ the `CreationService` removed. Ran some final tests and everything seems to work just fine :)

Let me know if anything else needs changing.

Comment (akhenry) [inline platform/commonUI/edit/test/actions/SaveAsActionSpec.js]: @BogdanAlexandru it confused us too, and we are working to remove that dependency, which we believe is superfluous.

**Comment by akhenry** [comment]
@BogdanAlexandru Thanks for the quick turnaround on this. I believe this is good to merge now. We are currently in a code freeze however, so I'm going to hold off until we're out of freeze.

Thanks again for your help!

**Comment by BogdanAlexandru** [comment]
Adding this here to remind me to update `NavigateAction` after https://github.com/nasa/openmct/pull/1201/commits/3548cde9c44076c687038f20ebeec77939fd91e7 goes in otherwise the app will blow up at runtime.

**Comment by akhenry** [comment]
A merge conflict was introduced by another PR which I resolved via the CLI. This branch has now been merged.
